### PR TITLE
Feature/userstory 11

### PR DESCRIPTION
**Pull request by AstridOttmann**
I refactored the speech synthesis a third time. I hope that the voices are available now. 
Furthermore I implemented the translation feature (US#10 and US#11). 
Styling is only preliminary.

### CODE REVIEW

File: components/Buttons/AddButton.js [hunk 1]
```diff
@@ -0,0 +1,24 @@
+import StyledButton from "./StyledButton";
+import SVGIcon from "../Icons/SVGIcon";
+
+export default function AddButton({ isAddMode, onClick }) {
+  return (
+    <>
```
Comment (F-Kirchhoff) [inline components/Buttons/AddButton.js]: This fragment is only needed if you want to return two or more elements in parallel. Here you only return the StyledButton so you can remove the fragment.

File: components/Buttons/AddButton.js [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+        {isAddMode ? "Close" : "Add"}
+        {!isAddMode && (
+          <SVGIcon
+            variant="plus"
+            width="2rem"
+            color="#04BF45"
+            aria-label="plus"
+          />
+        )}
+      </StyledButton>
```
Comment (F-Kirchhoff) [inline components/Buttons/AddButton.js]: This is a nice component, but AddButton, SearchButton and TranslateButton are very similar. At some point you might think about generalizing these tree buttons into your StyledButton component and pass down the text and Icons as children (you can always use the prop children to receive the nested elements of a component). But for now its fine!

File: components/Buttons/SearchButton.js [hunk 3]
```diff
@@ -0,0 +1,24 @@
+import StyledButton from "./StyledButton";
+import SVGIcon from "../Icons/SVGIcon";
+
+export default function SearchButton({ isSearchMode, onClick }) {
+  return (
+    <>
```
Comment (F-Kirchhoff) [inline components/Buttons/SearchButton.js]: framgent not needed

File: components/Buttons/TranslateButton.js [hunk 4]
```diff
@@ -0,0 +1,24 @@
+import StyledButton from "./StyledButton";
+import SVGIcon from "../Icons/SVGIcon";
+
+export default function TranslateButton({ isTranslateMode, onClick }) {
+  return (
+    <>
```
Comment (F-Kirchhoff) [inline components/Buttons/TranslateButton.js]: same here

File: components/Form/index.js [hunk 5]
```diff
@@ -1,15 +1,19 @@
...
 export default function Form({
   onSubmitEvent,
   onFirstInput,
   entry,
   isEditMode,
+  availableVoices,
 }) {
   const router = useRouter();
 
+  console.log("avail", availableVoices);
```
Comment (F-Kirchhoff) [inline components/Form/index.js]: please remove all logs before merging 🕵️

File: components/Form/index.js [hunk 6]
```diff
@@ -55,6 +60,20 @@ export default function Form({
...
+          aria-label="select-voice"
+          defaultValue={isEditMode && entry.voiceURI}
+        >
+          {storedVoices.map((voice) => {
+            return (
+              <option key={voice.name} value={voice.name}>
+                {voice.name} {voice.lang}
+              </option>
+            );
+          })}
```
Comment (F-Kirchhoff) [inline components/Form/index.js:73]: nice! I like it very much that you generate your options dynamically.

File: components/TranslationForm/index.js [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+          target_lang: data.target_lang,
+        }),
+        headers: {
+          "Content-Type": "application/json",
+        },
+      });
+
+      if (response.ok) {
+        const result = await response.json();
+        console.log("result", result);
```
Comment (F-Kirchhoff) [inline components/TranslationForm/index.js]: 🕵️

File: components/TranslationForm/index.js [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+
+export default function TranslationForm() {
+  const [translation, setTranslation] = useState("");
+  const [detectedLanguage, setDetectedLanguage] = useState("");
+
+  async function handleSubmit(event) {
+    event.preventDefault();
+    const formData = new FormData(event.target);
+    const data = Object.fromEntries(formData);
+    console.log("data", data);
```
Comment (F-Kirchhoff) [inline components/TranslationForm/index.js]: 🕵️

File: pages/_app.js [hunk 9]
```diff
@@ -6,7 +6,7 @@ import useVoices from "@/hooks/useVoices";
 
 export default function App({ Component, pageProps }) {
   const availableVoices = useVoices();
-  console.log("ava", availableVoices);
+  //console.log("ava", availableVoices);
```
Comment (F-Kirchhoff) [inline pages/_app.js]: please remove code instead of commenting it out.

File: pages/api/deepl.js [hunk 10]
```diff
@@ -0,0 +1,29 @@
+export default async function handler(request, response) {
+  console.log("query", request.query);
+  console.log("body", request.body);
```
Comment (F-Kirchhoff) [inline pages/api/deepl.js]: 🕵️

File: pages/api/deepl.js [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+      target_lang: request.body.target_lang,
+    }),
+    headers: {
+      "Content-Type": "application/json",
+      Authorization: `DeepL-Auth-Key ${process.env.DEEPL_AUTH_KEY}`,
+    },
+  });
+
+  // const requestedTranslation = await result.json();
+  // response.status(201).json({ requestedTranslation });
```
Comment (F-Kirchhoff) [inline pages/api/deepl.js]: here to

File: pages/index.js [hunk 12]
```diff
@@ -8,11 +8,17 @@ import Link from "next/link";
...
+import AddButton from "@/components/Buttons/AddButton";
+import SearchButton from "@/components/Buttons/SearchButton";
+import TranslateButton from "@/components/Buttons/TranslateButton";
 
 export default function HomePage() {
   const [translationList, setTranslationList] = useAtom(translationListAtom);
   const [isFound, setIsFound] = useState();
+  const [isAddMode, setIsAddMode] = useState(false);
+  const [isSearchMode, setIsSearchMode] = useState(false);
+  const [isTranslateMode, setIsTranslateMode] = useState(false);
```
Comment (F-Kirchhoff) [inline pages/index.js:21]: Are these modes independent from each other? In other words: can I open all Forms at the same time or should only be one form active at a time?

**Review by F-Kirchhoff — CHANGES_REQUESTED** [review]
Hey Astrid,
this looks amazing, you already implement a lot of best practices. I only found some minor issues. If you want to step up your game in the future you can look if you can generalize your button components. Very often we can use the same component for multiple purposes. Instead of thinking: this button is for toggling the "show Add form state" you should think: this button is for toggling "some state". Then you can toggle many different states with the same button component! The only thing that has to change is the onClick function you pass to that button.

File: components/Buttons/AddButton.js (hunk 1 above)
Comment (AstridOttmann) [reply components/Buttons/AddButton.js]: done!

File: components/Buttons/AddButton.js (hunk 2 above)
Comment (AstridOttmann) [reply components/Buttons/AddButton.js]: Thank you! You are right, good point! 
I'll try to generalize them later on.

File: components/Buttons/SearchButton.js (hunk 3 above)
Comment (AstridOttmann) [reply components/Buttons/SearchButton.js]: done!

File: components/Form/index.js (hunk 5 above)
Comment (AstridOttmann) [reply components/Form/index.js]: done!

File: pages/index.js (hunk 12 above)
Comment (AstridOttmann) [reply pages/index.js:21]: Yes, I want them to be independent from each other. I want to be able to open them individually. But it should not be excluded to have several forms opened at the same time.

File: pages/api/deepl.js [hunk 13]
```diff
@@ -0,0 +1,24 @@
...
+    try {
+      const requestedTranslation = await result.json();
+      response.status(201).json({ requestedTranslation });
+    } catch (error) {
+      console.error(error);
+      response.status(400).json({ error: error.message });
+    }
+  }
+  return response.status(405).json({ status: "Method not allowed" });
+}
```
Comment (RobinWitt) [inline pages/api/deepl.js:24]: Nice!

**Review by F-Kirchhoff — APPROVED** [review]
LGTM 🦊

**Comment by ghost** [comment]
lgtm

**Pull request merged**